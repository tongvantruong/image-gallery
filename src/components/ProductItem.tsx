import { Card, Typography, Image, Flex } from "antd";
import { useState } from "react";
import { createPortal } from "react-dom";
import { Product } from "@/types/product";

type ProductItemProps = Product;

const ProductItem = ({ id, title, thumbnail, images }: ProductItemProps) => {
  const [imagesToPreview, setImagesToPreview] = useState<string[]>([]);

  return (
    <Card style={{ margin: 20 }} hoverable key={id}>
      <Flex vertical align="center" justify="center">
        <Image
          alt="product thumbnail"
          data-testid="product-thumbnail"
          src={thumbnail}
          preview={{ visible: false }}
          onClick={() => {
            console.log("clicked");
            setImagesToPreview(images);
          }}
        />
        <Typography.Text data-testid="product-title">{title}</Typography.Text>
      </Flex>
      {imagesToPreview.length > 0
        ? createPortal(
            <Image.PreviewGroup
              preview={{
                visible: true,
                onVisibleChange: (value) => {
                  if (!value) setImagesToPreview([]);
                },
              }}
            >
              {imagesToPreview.map((image, index) => {
                return (
                  <Image
                    key={index}
                    src={image}
                    alt="product preview image"
                  ></Image>
                );
              })}
            </Image.PreviewGroup>,
            document.body
          )
        : null}
    </Card>
  );
};

export default ProductItem;
