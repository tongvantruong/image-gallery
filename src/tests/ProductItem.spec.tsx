import "@testing-library/jest-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import ProductItem from "@/components/ProductItem";
import { describe, expect, it } from "vitest";

describe("ProductItem", () => {
  it("should match snapshot", () => {
    const id = "test-id";
    const title = "test-title";
    const thumbnail = "test-thumbnail";
    const images = ["image1.png", "image2.png"];
    const { container } = render(
      <ProductItem
        id={id}
        title={title}
        thumbnail={thumbnail}
        images={images}
      />
    );

    expect(container).toMatchSnapshot();
  });
  it("should render title and thumbnail props if provided", () => {
    const id = "test-id";
    const title = "test-title";
    const thumbnail = "test-thumbnail";
    const images = ["image1.png", "image2.png"];
    render(
      <ProductItem
        id={id}
        title={title}
        thumbnail={thumbnail}
        images={images}
      />
    );

    expect(screen.getByTestId("product-title").innerHTML).toBe("test-title");
    expect(
      screen.getByAltText("product thumbnail").getAttribute("src")
    ).toContain("test-thumbnail");
  });
  it("should render Preview if clicked the thumbnail", async () => {
    const id = "test-id";
    const title = "test-title";
    const thumbnail = "test-thumbnail";
    const images = ["image1.png", "image2.png"];
    render(
      <ProductItem
        id={id}
        title={title}
        thumbnail={thumbnail}
        images={images}
      />
    );

    const thumbnailElement = screen.getByAltText("product thumbnail");
    await fireEvent.click(thumbnailElement);
    expect(screen.getAllByAltText("product preview image").length).toBe(2);
    expect(
      screen.getAllByAltText("product preview image").at(0)
    ).toBeInTheDocument();
    expect(
      screen.getAllByAltText("product preview image").at(1)
    ).toBeInTheDocument();
  });
});
