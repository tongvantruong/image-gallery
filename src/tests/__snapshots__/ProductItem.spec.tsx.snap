// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductItem > should match snapshot 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered ant-card-hoverable css-dev-only-do-not-override-161f05s"
    style="margin: 20px;"
  >
    <div
      class="ant-card-body"
    >
      <div
        class="ant-flex css-dev-only-do-not-override-161f05s ant-flex-align-center ant-flex-justify-center ant-flex-vertical"
      >
        <div
          class="ant-image css-dev-only-do-not-override-161f05s"
          data-testid="product-thumbnail"
        >
          <img
            alt="product thumbnail"
            class="ant-image-img css-dev-only-do-not-override-161f05s"
            src="test-thumbnail"
          />
          <div
            class="ant-image-mask"
          >
            <div
              class="ant-image-mask-info"
            >
              <span
                aria-label="eye"
                class="anticon anticon-eye"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="eye"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112 112 50.1 112 112-50.1 112-112 112z"
                  />
                </svg>
              </span>
              Preview
            </div>
          </div>
        </div>
        <span
          class="ant-typography css-dev-only-do-not-override-161f05s"
          data-testid="product-title"
        >
          test-title
        </span>
      </div>
    </div>
  </div>
</div>
`;
